### DISCUSSION TITLE
NPM Package missing?

### DISCUSSION THREAD

**Discussion by jhandke** [borkdude/sci#883] [Q&A]
Hello @borkdude, thanks for this amazing piece of software!
I would like to use sci to create an evaluator module for a basic Clojure programming environment for intro students based on Googles Blockly. Sadly the NPM Package is not available anymore.

Is there any chance that you could upload it again? It would be very helpful.

**Comment by borkdude** [comment]
Hi @jhandke - SCI isn't available on npm, but https://github.com/babashka/nbb is. You probably want to be using that for Node.js instead of SCI. Can you let me know if that works for you? I'd be happy to help you further.

**Reply by borkdude** [reply]
If you are looking for something purely client side: scittle may work.

https://github.com/babashka/scittle

**Reply by jhandke** [reply]
My project should evaluate Clojure in the browser so i wanted to use sci's Javascript API. The rest of my project is written in Javascript because integrating Blockly with ClojureScript is a lot of work and really just JS interop. So no real advantages of using ClojureScript. 
So for me the function evalString from sci would be perfect as I dont want to evaluate files but just small blocks of code.

Regardless of using sci or nbb: can I use it in the browser with something like browserify?

**Comment by jhandke** [comment]
Update: When adding sci as a submodule in my project and running /scripts/release-js i can do `import { evalString } from './sci/out/min/sci.min.js'` and make use of evalString.

**Comment by borkdude** [comment]
@jhandke The NPM package is no longer supported, so `script/release_js` isn't supposed to be used and will probably be removed.

Instead I recommend using [scittle](https://babashka.org/scittle/) for this purpose. Try this HTML page:

``` html
<html>
  <head>
    <script src="https://cdn.jsdelivr.net/npm/scittle@0.5.14/dist/scittle.js" type="application/javascript"></script>
  </head>
  <body>
    <input id="code" value="(js/alert (+ 1 2 3))">
    <button onclick="scittle.core.eval_string(document.querySelector('#code').value)">
      Eval!
    </button>
  </body>
</html>
```

When you click the `Eval!` button, you will see an alert with `6`.

**Reply by jhandke** [reply]
Thanks for this recommendation! 
Is scittle context-aware? I want to declare functions and call them in a later evaluation.

**Reply by borkdude** [reply]
Yes. Consider this HTML page:

``` html
<html>
  <head>
    <script src="https://cdn.jsdelivr.net/npm/scittle@0.5.14/dist/scittle.js" type="application/javascript"></script>
    <script type="application/x-scittle">
      (ns foo)
      (defn foobar [] 1337)
    </script>
  </head>
  <body>
    <input id="code" value="(js/alert (+ 1 2 3))">
    <button onclick="scittle.core.eval_string(document.querySelector('#code').value)">
      Eval!
    </button>
  </body>
</html>
```

Then when you enter the code `(js/alert (+ 1 2 (foo/foobar)))` it will pop up with `1340`.

**Reply by jhandke** [reply]
Thanks! This is working fine!
Now I have to find a way to import scittle as a ES6 module in my existing code.

**Reply by borkdude** [reply]
@jhandke Scittle itself isn't compiled as an ES6 module so you don't import it. Just use the global `scittle.core.eval_string` after loading the script tag.

**Reply by borkdude** [reply]
Nbb (not intended for browser) is compiled as an ES6 module though, if you want an example of SCI + es6:

https://github.com/babashka/nbb/blob/main/shadow-cljs.edn

**Reply by borkdude** [reply]
There is also an alternative available as an ES6 module, an experimental CLJS compiler I'm working on:

https://squint-cljs.github.io/cherry/

You can install it from npm:

```
npm install cherry-cljs
```

and then use it with:

``` clojure
import { compileString} from 'cherry-cljs';

const jsString = compileString('(prn (assoc {} :foo :bar))');
eval(jsString);
```

Note that cherry is currently more experimental than SCI (since SCI is older and was used more).

**Reply by jhandke** [reply]
Thank you so much for your help!
I will test all those options and report which one is most useful for me.
